library/std/src/io/stdio.rs

```diff
@@ -1017,6 +1018,123 @@ where
     }
 }
 
+mod terminal_sealed {
+    #[unstable(feature = "sealed", issue = "none")]
+    pub trait Sealed {}
+}
+
+/// Trait to determine if a descriptor/handle refers to a terminal/tty.
+#[unstable(feature = "is_terminal", issue = "98070")]
+pub trait IsTerminal: terminal_sealed::Sealed {
+    /// Returns true if the descriptor/handle refers to a terminal/tty.
+    ///
+    /// Returns an error if any errors occurred when attempting to make this determination, other
+    /// than those expected as part of making this determination. For instance, on some platforms
+    /// checking for a terminal may involve making a system call that returns an error on a
+    /// non-terminal, so that error will get translated to `Ok(false)`, but any unexpected error
+    /// will still get returned as an error.
+    ///
+    /// On platforms where Rust does not know how to detect a terminal yet, this will return
+    /// `Ok(false)`.
+    fn is_terminal(&self) -> io::Result<bool>;
+}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for Stdin {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for Stdout {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for Stderr {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for File {}
+
+cfg_if::cfg_if! {
+    if #[cfg(unix)] {
```

Comment:
According to a code comment in tidy there's the goal of moving platform-specific cfgs out of stdio.rs. So we shouldn't add more.

Reply:
I looked at doing this, and as far as I can tell, it'd be *much* less maintainable that way. It'd be easy enough to move the Windows and UNIX/WASI implementations, but rather than a simple `else` case, we'd need stubs in unsupported that would need a re-export for every OS. That seems like a substantial hit to maintainability.

I'm willing to do that, if libs feels like it's necessary, but in this context I think this is a case study in maintenance tradeoffs.

Reply:
I'd prefer to keep the platform-specific code in the `sys` and `os` modules as much as possible. In the future, we might want to try splitting off that part into a separate crate, or something that can be provided by an external crate through some future language mechanism. Having platform-specific code in the rest of `std` makes that step a lot harder.

Reply:
> but rather than a simple `else` case, we'd need stubs in unsupported that would need a re-export for every OS.

That's a good thing, right? Then we know exactly for which operating systems we don't have a useful IsTerminal implementation.

Reply:
I think this might work well:

- Have io::stdio implement IsTerminal only for the types that exist on all platforms, and have them all forward to `std::sys::is_terminal(self)`:

```rust
macro_rules! impl_for_asfd {
    ($($t:ty),*$(,)?) => {$(
        #[unstable(feature = "is_terminal", issue = "98070")]
        impl IsTerminal for $t {
            #[inline]
            fn is_terminal(&self) -> bool {
                crate::sys::io::is_terminal(self)
            }
        }
    )*}
}

impl_for_asfd!(File, Stdin, StdinLock<'_>, Stdout, StdoutLock<'_>, Stderr, StderrLock<'_>,);
```

- Add `is_terminal` functions with a simple `&impl SomeTrait` argument for each platform in their `sys::io` module:

```rust
// unix:
pub fn is_terminal(x: &impl AsFd) -> bool {
    unsafe { libc::isatty(x.as_fd().as_raw_fd()) != 0 }
}
```
```rust
// windows:
pub fn is_terminal(x: &impl AsHandle) {
    unsafe { handle_is_console(self.as_handle()) } // (Or just inline it.)
}
```
```rust
// unsupported:
pub fn is_terminal<T>(x: &T) {
    false
}
```
- Implement `IsTerminal` for `BorrowedFd` (etc.) directly in the place where those types are defined:
```rust
// library/std/src/os/fd/owned.rs
impl IsTerminal for OwnedFd {
    fn is_terminal(&self) -> bool {
        crate::sys::io::is_terminal(self)
    }
}

impl IsTerminal for BorrowedFd<'_> {
    fn is_terminal(&self) -> bool {
        crate::sys::io::is_terminal(self)
    }
}
```
(And similar in `library/std/src/os/windows/io/handle.rs`.)

---

That way, the public interface is defined outside sys, but the implementation details are inside of it. That makes it much harder to accidentally break an (untested) platform, or to accidentally change the public API on a platform when changing something about the implementation.

Reply:
@m-ou-se That's brilliant, thank you. Working on that now.